src/sentry/deletions/defaults/data_source.py

```diff
@@ -0,0 +1,27 @@
+from sentry.deletions.base import BaseRelation, ModelDeletionTask, ModelRelation
+from sentry.workflow_engine.models.data_source import DataSource
+
+
+class DataSourceDeletionTask(ModelDeletionTask[DataSource]):
+    def get_child_relations(self, instance: DataSource) -> list[BaseRelation]:
+        from sentry.snuba.models import QuerySubscription, SnubaQuery
+
+        model_relations: list[BaseRelation] = [
+            ModelRelation(QuerySubscription, {"id": instance.query_id})
+        ]
```

Comment:
Do we need to check the type here? `query_id` will point to different datasources in the future.

Looks like you can do `instance.type_handler.bulk_get_query_object([instance])` to get all related objects.

I think that probably, you should build the `SnubaQuery` deletion stuff into a delete module for `subscriptions` rather than here. That way, you can just add the `QuerySubscription` as the model relation, and then it will automatically clean itself up.

Reply:
Today we only appear to have a single `type` for a `QuerySubscription` (`incident`). Since I moved deleting the snuba query to the query subscription deletion task do I still need to filter?

I moved deleting it into `querysubscription.py` but ended up having to check for an `AlertRule` first - after everything is migrated that won't be a problem and I'm not sure if in practice it'll ever happen that we're deleting a `QuerySubscription` that still has an active `AlertRule`.

Reply:
I think you still need to use `bulk_get_query_object` here or the `QuerySubscription` rows won't be deleted when the `DataSource` is deleted. Probably you'd need to schedule deletions for them, since we don't have an explicit relationship that we can follow here...